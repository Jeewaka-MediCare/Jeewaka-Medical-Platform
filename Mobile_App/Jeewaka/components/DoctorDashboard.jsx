import React, { useState, useEffect } from 'react';
import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
  Image,
  RefreshControl,
  Alert,
  Animated,
  Dimensions
} from 'react-native';
import { Ionicons, MaterialCommunityIcons } from '@expo/vector-icons';
import { useRouter } from 'expo-router';
import { format, parseISO } from 'date-fns';
import api from '../services/api';
import useAuthStore from '../store/authStore';

export default function DoctorDashboard() {
  const { user } = useAuthStore();
  const router = useRouter();
  const [doctorData, setDoctorData] = useState(null);
  const [ratingSummary, setRatingSummary] = useState(null);
  const [recentReviews, setRecentReviews] = useState([]);
  const [loading, setLoading] = useState(true);
  const [refreshing, setRefreshing] = useState(false);
  
  // Animation values for floating reviews
  const [fadeAnim] = useState(new Animated.Value(0));
  const [slideAnim] = useState(new Animated.Value(50)); // Start 50 pixels below
  const [scrollAnim] = useState(new Animated.Value(0)); // For continuous scrolling animation

  // Mock reviews for testing
  const mockReviews = [
    {
      patient: { name: 'John Doe' },
      rating: 5,
      comment: 'Excellent doctor! Very professional and caring. Highly recommend for orthopedic treatments.',
      createdAt: new Date().toISOString()
    },
    {
      patient: { name: 'Jane Smith' },
      rating: 4,
      comment: 'Great experience. Dr. Priya explained everything clearly and the treatment was effective.',
      createdAt: new Date(Date.now() - 86400000).toISOString() // 1 day ago
    }
  ];

  // Function to start continuous scrolling animation
  const startContinuousScroll = (reviewCount) => {
    console.log('Starting continuous scroll for', reviewCount, 'reviews');
    if (reviewCount === 0) return;
    
    const cardWidth = 290; // Card width (280) + margin (10)
    const totalWidth = cardWidth * reviewCount;
    
    const scrollSequence = () => {
      scrollAnim.setValue(0);
      console.log('Scrolling to:', -totalWidth);
      Animated.timing(scrollAnim, {
        toValue: -totalWidth,
        duration: reviewCount * 4000, // 4 seconds per card
        useNativeDriver: true,
      }).start(() => {
        // Reset and restart the animation
        console.log('Animation completed, restarting...');
        scrollSequence();
      });
    };
    
    scrollSequence();
  };

  const fetchDoctorData = async () => {
    if (!user?._id) return;
    
    try {
      console.log('Fetching doctor dashboard data for:', user._id);
      const { data } = await api.get(`/api/doctorCard/${user._id}`);
      setDoctorData(data.doctor);
      setRatingSummary(data.ratingSummary);
      
      // Fetch recent reviews
      try {
        const reviewsResponse = await api.get(`/api/ratings/doctor/${user._id}`);
        const reviews = reviewsResponse.data || [];
        // Get the 10 most recent reviews
        const sortedReviews = reviews
          .sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt))
          .slice(0, 10);
        setRecentReviews(sortedReviews);
        console.log('Reviews set:', sortedReviews.length);
        
        // For testing: Always use mock reviews to ensure animation works
        // Comment out this section when you have real reviews
        if (sortedReviews.length === 0) {
          setRecentReviews(mockReviews);
          console.log('Set mock reviews for testing');
        }
      } catch (reviewError) {
        console.error('Failed to fetch reviews:', reviewError);
        // Set mock reviews for testing even on error
        setRecentReviews(mockReviews);
      }
    } catch (error) {
      console.error('Failed to fetch doctor data:', error);
      Alert.alert('Error', 'Failed to load your profile data');
    } finally {
      setLoading(false);
      setRefreshing(false);
    }
  };

  useEffect(() => {
    fetchDoctorData();
  }, [user?._id]);

  // Separate useEffect to handle animations when reviews change
  useEffect(() => {
    if (recentReviews.length > 0) {
      // Start animation when reviews are available
      console.log('Starting reviews section animation for', recentReviews.length, 'reviews');
      
      // Reset animation values
      fadeAnim.setValue(0);
      slideAnim.setValue(50);
      
      setTimeout(() => {
        Animated.parallel([
          Animated.timing(fadeAnim, {
            toValue: 1,
            duration: 800,
            useNativeDriver: true,
          }),
          Animated.spring(slideAnim, {
            toValue: 0,
            tension: 100,
            friction: 8,
            useNativeDriver: true,
          }),
        ]).start(() => {
          console.log('Animation completed');
          // Start continuous scrolling after fade-in completes
          setTimeout(() => {
            startContinuousScroll(recentReviews.length);
          }, 2000); // Wait 2 seconds before starting scroll
        });
      }, 1000);
    }
  }, [recentReviews]);

  const onRefresh = () => {
    setRefreshing(true);
    fetchDoctorData();
  };

  const handleViewReviews = () => {
    if (doctorData?._id) {
      // Navigate to dedicated reviews page
      router.push({
        pathname: '/doctor-reviews',
        params: { doctorId: doctorData._id }
      });
    }
  };

  const getGreeting = () => {
    const hour = new Date().getHours();
    if (hour < 12) return 'Good Morning';
    if (hour < 17) return 'Good Afternoon';
    return 'Good Evening';
  };

  if (loading && !doctorData) {
    return (
      <View style={styles.loadingContainer}>
        <Text style={styles.loadingText}>Loading your dashboard...</Text>
      </View>
    );
  }

  const avgRating = ratingSummary?.avgRating || 0;
  const totalReviews = ratingSummary?.totalReviews || 0;

  return (
    <View style={styles.container}>
      <ScrollView 
        style={styles.scrollView}
        refreshControl={
          <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
        }
      >
        {/* Welcome Section */}
        <View style={styles.welcomeSection}>
          <View style={styles.welcomeContent}>
            <Text style={styles.greeting}>{getGreeting()},</Text>
            <Text style={styles.doctorName}>{doctorData?.name || 'Doctor'}</Text>
            <Text style={styles.welcomeMessage}>Welcome to your dashboard</Text>
          </View>
          <Image 
            source={
              doctorData?.profile 
                ? { uri: doctorData.profile } 
                : require('../assets/images/doctor-placeholder.png')
            } 
            style={styles.profileImage} 
          />
      </View>

      {/* Quick Stats */}
      <View style={styles.statsContainer}>
        <View style={styles.statCard}>
          <MaterialCommunityIcons name="star" size={24} color="#FFD700" />
          <Text style={styles.statValue}>{avgRating.toFixed(1)}</Text>
          <Text style={styles.statLabel}>Rating</Text>
        </View>
        <View style={styles.statCard}>
          <MaterialCommunityIcons name="comment-multiple" size={24} color="#2563EB" />
          <Text style={styles.statValue}>{totalReviews}</Text>
          <Text style={styles.statLabel}>Reviews</Text>
        </View>
        <View style={styles.statCard}>
          <MaterialCommunityIcons name="clock" size={24} color="#10B981" />
          <Text style={styles.statValue}>{doctorData?.yearsOfExperience || 0}</Text>
          <Text style={styles.statLabel}>Years Exp</Text>
        </View>
      </View>

      {/* Recent Reviews Section */}
      <Animated.View 
        style={[
          styles.reviewsSection,
          {
            opacity: fadeAnim,
            transform: [{ translateY: slideAnim }]
          }
        ]}
      >
        <View style={styles.reviewsSectionHeader}>
          <MaterialCommunityIcons name="star" size={24} color="#FFD700" />
          <Text style={styles.reviewsSectionTitle}>Recent Reviews</Text>
          <TouchableOpacity onPress={handleViewReviews}>
            <Text style={styles.viewAllReviewsButton}>View All</Text>
          </TouchableOpacity>
        </View>
        
        {recentReviews.length > 0 ? (
          <View style={styles.reviewsContainer}>
            <Animated.View
              style={[
                styles.reviewsAnimatedContainer,
                {
                  transform: [{ translateX: scrollAnim }]
                }
              ]}
            >
              {/* Render reviews twice for seamless loop */}
              {[...recentReviews, ...recentReviews].map((review, index) => (
                <Animated.View 
                  key={`${review.id || index}-${index >= recentReviews.length ? 'duplicate' : 'original'}`}
                  style={[
                    styles.reviewCard,
                    {
                      opacity: fadeAnim,
                      transform: [{ 
                        scale: fadeAnim.interpolate({
                          inputRange: [0, 1],
                          outputRange: [0.8, 1]
                        })
                      }]
                    }
                  ]}
                >
                  <View style={styles.reviewCardHeader}>
                    <Text style={styles.reviewerName} numberOfLines={1}>
                      {review.patient?.name || 'Anonymous'}
                    </Text>
                    <View style={styles.reviewStars}>
                      {[1, 2, 3, 4, 5].map((star) => (
                        <Ionicons
                          key={star}
                          name="star"
                          size={14}
                          color={star <= review.rating ? "#FFD700" : "#CBD5E1"}
                        />
                      ))}
                    </View>
                  </View>
                  <Text style={styles.reviewComment} numberOfLines={4}>
                    {review.comment}
                  </Text>
                  <Text style={styles.reviewDate}>
                    {review.createdAt ? format(parseISO(review.createdAt), 'MMM dd, yyyy') : 'Recent'}
                  </Text>
                </Animated.View>
              ))}
            </Animated.View>
          </View>
        ) : (
          <View style={styles.noReviewsContainer}>
            <MaterialCommunityIcons name="star-outline" size={48} color="#CBD5E1" />
            <Text style={styles.noReviewsText}>No reviews yet</Text>
            <Text style={styles.noReviewsSubtext}>
              Your first patient reviews will appear here
            </Text>
          </View>
        )}
      </Animated.View>

      {/* Professional Information */}
      <View style={styles.infoSection}>
        <Text style={styles.sectionTitle}>Professional Information</Text>
        
        <View style={styles.infoCard}>
          <View style={styles.infoRow}>
            <MaterialCommunityIcons name="medical-bag" size={20} color="#2563EB" />
            <Text style={styles.infoLabel}>Specialization</Text>
          </View>
          <Text style={styles.infoValue}>
            {doctorData?.specialization || 'Not specified'}
          </Text>
          {doctorData?.subSpecializations && doctorData.subSpecializations.length > 0 && (
            <Text style={styles.subSpecText}>
              {doctorData.subSpecializations.join(', ')}
            </Text>
          )}
        </View>

        <View style={styles.infoCard}>
          <View style={styles.infoRow}>
            <MaterialCommunityIcons name="certificate" size={20} color="#2563EB" />
            <Text style={styles.infoLabel}>Qualifications</Text>
          </View>
          <Text style={styles.infoValue}>
            {doctorData?.qualifications && doctorData.qualifications.length > 0 
              ? doctorData.qualifications.join(', ') 
              : 'Not specified'
            }
          </Text>
        </View>

        <View style={styles.infoCard}>
          <View style={styles.infoRow}>
            <MaterialCommunityIcons name="card-account-details" size={20} color="#2563EB" />
            <Text style={styles.infoLabel}>Registration Number</Text>
          </View>
          <Text style={styles.infoValue}>
            {doctorData?.regNo || 'Not specified'}
          </Text>
        </View>

        <View style={styles.infoCard}>
          <View style={styles.infoRow}>
            <MaterialCommunityIcons name="translate" size={20} color="#2563EB" />
            <Text style={styles.infoLabel}>Languages</Text>
          </View>
          <Text style={styles.infoValue}>
            {doctorData?.languagesSpoken && doctorData.languagesSpoken.length > 0 
              ? doctorData.languagesSpoken.join(', ') 
              : 'Not specified'
            }
          </Text>
        </View>

        <View style={styles.infoCard}>
          <View style={styles.infoRow}>
            <MaterialCommunityIcons name="cash" size={20} color="#2563EB" />
            <Text style={styles.infoLabel}>Consultation Fee</Text>
          </View>
          <Text style={styles.infoValue}>
            LKR {doctorData?.consultationFee || 0}
          </Text>
        </View>
      </View>

      {/* About Section */}
      {doctorData?.bio && doctorData.bio.trim() && (
        <View style={styles.infoSection}>
          <Text style={styles.sectionTitle}>About</Text>
          <View style={styles.bioCard}>
            <Text style={styles.bioText}>{doctorData.bio}</Text>
          </View>
        </View>
      )}

      {/* Bottom spacing */}
      <View style={styles.bottomSpacing} />
    </ScrollView>
  </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#F8FAFC',
  },
  scrollView: {
    flex: 1,
  },
  loadingContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#F8FAFC',
  },
  loadingText: {
    fontSize: 16,
    color: '#64748B',
  },
  welcomeSection: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 20,
    backgroundColor: 'white',
    marginBottom: 16,
    borderBottomWidth: 1,
    borderBottomColor: '#E2E8F0',
  },
  welcomeContent: {
    flex: 1,
  },
  greeting: {
    fontSize: 16,
    color: '#64748B',
    marginBottom: 4,
  },
  doctorName: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#1E293B',
    marginBottom: 4,
  },
  welcomeMessage: {
    fontSize: 14,
    color: '#64748B',
  },
  profileImage: {
    width: 80,
    height: 80,
    borderRadius: 40,
    borderWidth: 3,
    borderColor: '#E2E8F0',
  },
  statsContainer: {
    flexDirection: 'row',
    paddingHorizontal: 16,
    marginBottom: 16,
    gap: 12,
  },
  statCard: {
    flex: 1,
    backgroundColor: 'white',
    padding: 16,
    borderRadius: 12,
    alignItems: 'center',
    borderWidth: 1,
    borderColor: '#E2E8F0',
  },
  statValue: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#1E293B',
    marginTop: 8,
    marginBottom: 4,
  },
  statLabel: {
    fontSize: 12,
    color: '#64748B',
    textAlign: 'center',
  },
  infoSection: {
    paddingHorizontal: 16,
    marginBottom: 24,
  },
  sectionTitle: {
    fontSize: 20,
    fontWeight: 'bold',
    color: '#1E293B',
    marginBottom: 16,
  },
  infoCard: {
    backgroundColor: 'white',
    padding: 16,
    borderRadius: 12,
    marginBottom: 12,
    borderWidth: 1,
    borderColor: '#E2E8F0',
  },
  infoRow: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 8,
  },
  infoLabel: {
    fontSize: 14,
    fontWeight: '500',
    color: '#64748B',
    marginLeft: 12,
  },
  infoValue: {
    fontSize: 16,
    color: '#1E293B',
    lineHeight: 22,
  },
  subSpecText: {
    fontSize: 14,
    color: '#64748B',
    marginTop: 4,
    fontStyle: 'italic',
  },
  bioCard: {
    backgroundColor: 'white',
    padding: 16,
    borderRadius: 12,
    borderWidth: 1,
    borderColor: '#E2E8F0',
  },
  bioText: {
    fontSize: 15,
    color: '#334155',
    lineHeight: 22,
  },
  bottomSpacing: {
    height: 20,
  },
  // Recent Reviews Section Styles
  reviewsSection: {
    paddingHorizontal: 16,
    marginBottom: 24,
  },
  reviewsSectionHeader: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginBottom: 16,
  },
  reviewsSectionTitle: {
    flex: 1,
    fontSize: 20,
    fontWeight: 'bold',
    color: '#1E293B',
    marginLeft: 12,
  },
  viewAllReviewsButton: {
    fontSize: 14,
    color: '#2563EB',
    fontWeight: '600',
    paddingHorizontal: 12,
    paddingVertical: 6,
    backgroundColor: '#EFF6FF',
    borderRadius: 8,
    overflow: 'hidden',
  },
  reviewsHorizontalScroll: {
    marginHorizontal: -16,
  },
  reviewsScrollContent: {
    paddingHorizontal: 16,
  },
  reviewsContainer: {
    height: 200,
    overflow: 'hidden',
    marginHorizontal: -16,
  },
  reviewsAnimatedContainer: {
    flexDirection: 'row',
    paddingHorizontal: 16,
  },
  reviewCard: {
    width: 280,
    backgroundColor: 'white',
    borderRadius: 12,
    padding: 16,
    marginRight: 10,
    borderWidth: 1,
    borderColor: '#E2E8F0',
    elevation: 2,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 4,
  },
  reviewCardHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 12,
  },
  reviewerName: {
    fontSize: 16,
    fontWeight: '600',
    color: '#1E293B',
    flex: 1,
    marginRight: 8,
  },
  reviewStars: {
    flexDirection: 'row',
  },
  reviewComment: {
    fontSize: 14,
    color: '#64748B',
    lineHeight: 20,
    marginBottom: 12,
    minHeight: 60,
  },
  reviewDate: {
    fontSize: 12,
    color: '#9CA3AF',
    fontWeight: '500',
  },
  noReviewsContainer: {
    alignItems: 'center',
    paddingVertical: 40,
    paddingHorizontal: 20,
  },
  noReviewsText: {
    fontSize: 16,
    fontWeight: '600',
    color: '#64748B',
    marginTop: 12,
    marginBottom: 4,
  },
  noReviewsSubtext: {
    fontSize: 14,
    color: '#9CA3AF',
    textAlign: 'center',
  },
});