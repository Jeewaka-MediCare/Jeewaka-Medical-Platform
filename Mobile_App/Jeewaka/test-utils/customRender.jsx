import React from 'react';
import { render } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';

// Custom render function that includes providers
export const renderWithProviders = (ui, options = {}) => {
  const Wrapper = ({ children }) => {
    return (
      <NavigationContainer>
        {children}
      </NavigationContainer>
    );
  };

  return render(ui, { wrapper: Wrapper, ...options });
};

// Re-export everything
export * from '@testing-library/react-native';

// Override render method
export { renderWithProviders as render };