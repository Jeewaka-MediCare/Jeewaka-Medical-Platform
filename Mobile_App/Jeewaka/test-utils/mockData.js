// Mock data for testing

export const mockDoctor = {
  _id: '507f1f77bcf86cd799439011',
  uid: 'doctor-uid-123',
  name: 'Dr. John Smith',
  email: 'john.smith@example.com',
  specialization: 'Cardiology',
  subSpecializations: ['Interventional Cardiology', 'Heart Surgery'],
  yearsOfExperience: 15,
  qualifications: ['MBBS', 'MD Cardiology', 'FRCPC'],
  hospitalAffiliation: 'General Hospital',
  consultationFee: 150,
  avgRating: 4.5,
  totalRatings: 120,
  totalReviews: 120,
  profilePicture: 'https://example.com/doctor.jpg',
  bio: 'Experienced cardiologist with 15 years of practice.',
  isAvailable: true,
  gender: 'Male',
  languages: ['English', 'Spanish'],
  phoneNumber: '+1234567890',
  address: '123 Medical St, City, State 12345',
  dateOfBirth: '1975-05-15',
  licenseNumber: 'MD12345',
  isVerified: true,
  createdAt: '2023-01-01T00:00:00.000Z',
  updatedAt: '2024-01-01T00:00:00.000Z',
};

export const mockPatient = {
  _id: '507f1f77bcf86cd799439012',
  uid: 'patient-uid-123',
  name: 'Jane Doe',
  email: 'jane.doe@example.com',
  phoneNumber: '+1234567891',
  dateOfBirth: '1990-03-20',
  gender: 'Female',
  address: '456 Patient Ave, City, State 12345',
  emergencyContact: {
    name: 'John Doe',
    relationship: 'Spouse',
    phoneNumber: '+1234567892',
  },
  medicalHistory: ['Diabetes', 'Hypertension'],
  allergies: ['Penicillin'],
  currentMedications: ['Metformin', 'Lisinopril'],
  insuranceInfo: {
    provider: 'Health Insurance Co',
    policyNumber: 'POL123456',
  },
  createdAt: '2023-01-01T00:00:00.000Z',
  updatedAt: '2024-01-01T00:00:00.000Z',
};

export const mockAppointment = {
  _id: '507f1f77bcf86cd799439013',
  patientId: mockPatient._id,
  doctorId: mockDoctor._id,
  date: '2024-02-15',
  startTime: '10:00 AM',
  endTime: '10:30 AM', 
  duration: 30,
  type: 'consultation',
  status: 'confirmed',
  consultationFee: 150,
  paymentStatus: 'paid',
  notes: 'Regular checkup',
  sessionId: 'session-123',
  doctorName: mockDoctor.name,
  doctorSpecialization: mockDoctor.specialization,
  patientName: mockPatient.name,
  isOngoing: false,
  createdAt: '2024-01-01T00:00:00.000Z',
  updatedAt: '2024-01-01T00:00:00.000Z',
};

export const mockPayment = {
  _id: '507f1f77bcf86cd799439014',
  appointmentId: mockAppointment._id,
  patientId: mockPatient._id,
  doctorId: mockDoctor._id,
  amount: 150,
  currency: 'LKR',
  paymentMethod: 'stripe',
  stripePaymentIntentId: 'pi_test_123',
  status: 'completed',
  doctorName: mockDoctor.name,
  doctorSpecialization: mockDoctor.specialization,
  date: '2024-01-01',
  appointmentDate: '2024-02-15',
  type: 'consultation',
  paidAt: '2024-01-01T12:00:00.000Z',
  createdAt: '2024-01-01T12:00:00.000Z',
};

export const mockReview = {
  _id: '507f1f77bcf86cd799439015',
  patientId: mockPatient._id,
  doctorId: mockDoctor._id,
  appointmentId: mockAppointment._id,
  rating: 5,
  comment: 'Excellent doctor, very professional and caring.',
  createdAt: '2024-01-02T00:00:00.000Z',
};

export const mockDoctorsList = [
  mockDoctor,
  {
    ...mockDoctor,
    _id: '507f1f77bcf86cd799439016',
    uid: 'doctor-uid-456',
    name: 'Dr. Sarah Johnson',
    email: 'sarah.johnson@example.com',
    specialization: 'Dermatology',
    subSpecializations: ['Cosmetic Dermatology', 'Pediatric Dermatology'],
    consultationFee: 120,
    avgRating: 4.8,
    totalRatings: 95,
    totalReviews: 95, // Add both fields for compatibility
  },
  {
    ...mockDoctor,
    _id: '507f1f77bcf86cd799439017',
    uid: 'doctor-uid-789',
    name: 'Dr. Michael Brown',
    email: 'michael.brown@example.com',
    specialization: 'Pediatrics',
    subSpecializations: ['Neonatology', 'Pediatric Cardiology'],
    consultationFee: 100,
    avgRating: 4.6,
    totalRatings: 87,
    totalReviews: 87, // Add both fields for compatibility
  },
];

export const mockApiResponse = {
  success: true,
  data: mockDoctorsList,
  message: 'Doctors fetched successfully',
};

export const mockApiError = {
  success: false,
  error: 'Network error',
  message: 'Failed to fetch data',
};

export const mockAuthUser = {
  uid: 'user-uid-123',
  email: 'test@example.com',
  displayName: 'Test User',
  emailVerified: true,
  getIdToken: jest.fn(() => Promise.resolve('mock-token')),
  getIdTokenResult: jest.fn(() => Promise.resolve({
    claims: { role: 'patient' }
  })),
};

export const mockMedicalRecord = {
  _id: '507f1f77bcf86cd799439018',
  patientId: mockPatient._id,
  doctorId: mockDoctor._id,
  appointmentId: mockAppointment._id,
  title: 'Cardiology Consultation',
  diagnosis: 'Hypertension',
  treatment: 'Prescribed medication and lifestyle changes',
  medications: ['Lisinopril 10mg', 'Amlodipine 5mg'],
  notes: 'Patient shows good response to treatment',
  labResults: 'Blood pressure: 130/80 mmHg',
  attachments: ['test-result.pdf'],
  date: '2024-02-15',
  createdAt: '2024-02-15T10:30:00.000Z',
  updatedAt: '2024-02-15T10:30:00.000Z',
};

export const mockVideoCallSession = {
  sessionId: 'session-123',
  participantId: 'participant-456',
  meetingId: 'meeting-789',
  status: 'ready',
  isConnected: false,
  localParticipant: null,
  participants: [],
  error: null,
};

export const mockAppointmentsList = [
  mockAppointment,
  {
    ...mockAppointment,
    _id: '507f1f77bcf86cd799439019',
    date: '2024-03-01',
    startTime: '2:00 PM',
    endTime: '2:30 PM',
    status: 'pending',
    paymentStatus: 'pending',
    isOngoing: false,
  },
  {
    ...mockAppointment,
    _id: '507f1f77bcf86cd799439020',
    date: '2024-01-15',
    startTime: '11:00 AM',
    endTime: '11:30 AM',
    status: 'completed',
    paymentStatus: 'paid',
    isOngoing: false,
  },
];

export const mockNavigationProps = {
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
    reset: jest.fn(),
    setOptions: jest.fn(),
  },
  route: {
    params: {},
  },
};