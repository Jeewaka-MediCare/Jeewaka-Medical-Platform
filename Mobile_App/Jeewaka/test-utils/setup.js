import React from 'react';
import '@testing-library/jest-native/extend-expect';
import 'react-native-gesture-handler/jestSetup';

// Mock Firebase Auth - used by components for authentication
jest.mock('firebase/auth', () => ({
  getAuth: jest.fn(() => ({
    currentUser: null,
    onAuthStateChanged: jest.fn(),
  })),
  getReactNativePersistence: jest.fn(() => 'mockPersistence'),
  initializeAuth: jest.fn(() => ({
    currentUser: null,
    onAuthStateChanged: jest.fn(),
  })),
  onAuthStateChanged: jest.fn(),
  signInWithEmailAndPassword: jest.fn(),
  createUserWithEmailAndPassword: jest.fn(),
  signOut: jest.fn(),
}));

// Mock Firebase App
jest.mock('firebase/app', () => ({
  initializeApp: jest.fn(() => ({})),
  getApps: jest.fn(() => []),
  getApp: jest.fn(() => ({})),
}));

// Mock react-native-gifted-charts - used by components for charts
jest.mock('react-native-gifted-charts', () => ({
  LineChart: jest.fn(({ children, ...props }) => null),
  BarChart: jest.fn(({ children, ...props }) => null),
  PieChart: jest.fn(({ children, ...props }) => null),
}));

// Mock Expo modules
jest.mock('expo-router', () => ({
  useRouter: () => ({
    push: jest.fn(),
    replace: jest.fn(),
    back: jest.fn(),
    canGoBack: jest.fn(() => true),
  }),
  useLocalSearchParams: () => ({}),
  router: {
    push: jest.fn(),
    replace: jest.fn(),
    back: jest.fn(),
  },
  Stack: {
    Screen: ({ children, ...props }) => children || null,
  },
  Tabs: {
    Screen: ({ children, ...props }) => children || null,
  },
}));

jest.mock('expo-font', () => ({
  useFonts: () => [true],
}));

jest.mock('expo-status-bar', () => ({
  StatusBar: 'StatusBar',
}));

jest.mock('expo-splash-screen', () => ({
  hideAsync: jest.fn(),
  preventAutoHideAsync: jest.fn(),
}));

jest.mock('expo-constants', () => ({
  default: {
    expoConfig: {
      extra: {},
    },
  },
}));

// Mock AsyncStorage
jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve()),
  removeItem: jest.fn(() => Promise.resolve()),
  clear: jest.fn(() => Promise.resolve()),
}));

// Mock Reanimated
jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');
  Reanimated.default.call = () => {};
  return Reanimated;
});

// Mock Safe Area Context
jest.mock('react-native-safe-area-context', () => ({
  SafeAreaProvider: ({ children }) => children,
  SafeAreaView: ({ children }) => children,
  useSafeAreaInsets: () => ({ top: 0, right: 0, bottom: 0, left: 0 }),
}));

// Mock Vector Icons - this handles all icon families used by components
jest.mock('@expo/vector-icons', () => {
  const React = require('react');
  const { Text } = require('react-native');
  
  const createMockIcon = (iconFamily) => {
    const MockIcon = (props) => 
      React.createElement(Text, { testID: `${iconFamily}-${props.name || 'icon'}` }, props.name || 'Icon');
    MockIcon.displayName = iconFamily;
    return MockIcon;
  };

  return {
    Ionicons: createMockIcon('Ionicons'),
    MaterialCommunityIcons: createMockIcon('MaterialCommunityIcons'),
    MaterialIcons: createMockIcon('MaterialIcons'),
    Feather: createMockIcon('Feather'),
    FontAwesome: createMockIcon('FontAwesome'),
    FontAwesome5: createMockIcon('FontAwesome5'),
    AntDesign: createMockIcon('AntDesign'),
    Entypo: createMockIcon('Entypo'),
    EvilIcons: createMockIcon('EvilIcons'),
    Foundation: createMockIcon('Foundation'),
    Octicons: createMockIcon('Octicons'),
    SimpleLineIcons: createMockIcon('SimpleLineIcons'),
    Zocial: createMockIcon('Zocial'),
  };
});

// Silence the warning about act() not being wrapped
const originalConsoleError = console.error;
beforeAll(() => {
  console.error = (...args) => {
    if (
      typeof args[0] === 'string' &&
      args[0].includes('Warning: ReactDOM.render is no longer supported')
    ) {
      return;
    }
    originalConsoleError.call(console, ...args);
  };
});

afterAll(() => {
  console.error = originalConsoleError;
});

// Global test timeout
jest.setTimeout(10000);