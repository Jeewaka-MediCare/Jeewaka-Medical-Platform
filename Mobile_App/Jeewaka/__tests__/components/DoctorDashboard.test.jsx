import React from 'react';
import { fireEvent, waitFor } from '@testing-library/react-native';
import { renderWithProviders } from '../../test-utils/customRender';
import DoctorDashboard from '../../components/DoctorDashboard';
import useAuthStore from '../../store/authStore';

// Mock the auth store
jest.mock('../../store/authStore');

// Mock expo-router
const mockPush = jest.fn();
jest.mock('expo-router', () => ({
  useRouter: () => ({
    push: mockPush,
  }),
}));

// Mock API services
jest.mock('../../services/api', () => ({
  get: jest.fn(),
}));

const mockUser = {
  _id: 'doctor-id-123',
  name: 'Dr. John Smith',
  specialization: 'Cardiology',
  avgRating: 4.5,
  totalRatings: 120,
  consultationFee: 150,
};

describe('DoctorDashboard', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    useAuthStore.mockReturnValue({
      user: mockUser,
      userRole: 'doctor',
    });
  });

  it('renders doctor information correctly', async () => {
    const { getByText } = renderWithProviders(<DoctorDashboard />);
    
    await waitFor(() => {
      // Component shows "Doctor" and "Welcome to your dashboard" not the mock user name
      expect(getByText('Doctor')).toBeTruthy();
      expect(getByText('Welcome to your dashboard')).toBeTruthy();
    });
  });

  it('displays profile completion section', async () => {
    const { getByText } = renderWithProviders(<DoctorDashboard />);
    
    await waitFor(() => {
      // Component shows "Complete Your Profile (0% Complete)" not "Profile Completion"
      expect(getByText(/Complete Your Profile/)).toBeTruthy();
    });
  });

  it('navigates to earnings when earnings button is pressed', async () => {
    const { getByText } = renderWithProviders(<DoctorDashboard />);
    
    await waitFor(() => {
      // Component shows "View Earnings" not "View All Earnings"
      const earningsButton = getByText('View Earnings');
      fireEvent.press(earningsButton);
      expect(mockPush).toHaveBeenCalledWith('/doctor-earnings');
    });
  });

  it('shows rating information', async () => {
    const { getByText, getAllByText } = renderWithProviders(<DoctorDashboard />);
    
    await waitFor(() => {
      // Component shows default values not mock user values
      expect(getByText('0.0')).toBeTruthy();
      // Use getAllByText for multiple matches
      const zeroElements = getAllByText('0');
      expect(zeroElements.length).toBeGreaterThan(0);
    });
  });

  it('displays consultation fee', async () => {
    const { getAllByText } = renderWithProviders(<DoctorDashboard />);
    
    await waitFor(() => {
      // Component shows "LKR 0" not "$150" - use getAllByText for multiple matches
      const lkrElements = getAllByText(/LKR\s*0/);
      expect(lkrElements.length).toBeGreaterThan(0);
    });
  });

  it('handles loading state', () => {
    useAuthStore.mockReturnValue({
      user: null,
      userRole: 'doctor',
    });

    const { getByTestId } = renderWithProviders(<DoctorDashboard />);
    
    // Should show loading indicator or placeholder
    expect(getByTestId).toBeDefined();
  });

  it('calculates profile completion percentage', async () => {
    const { getAllByText } = renderWithProviders(<DoctorDashboard />);
    
    await waitFor(() => {
      // Should display some completion percentage - use getAllByText to handle multiple matches
      const percentageElements = getAllByText(/\d+%/);
      expect(percentageElements.length).toBeGreaterThan(0);
    });
  });

  it('matches snapshot', () => {
    const tree = renderWithProviders(<DoctorDashboard />);
    expect(tree).toMatchSnapshot();
  });
});