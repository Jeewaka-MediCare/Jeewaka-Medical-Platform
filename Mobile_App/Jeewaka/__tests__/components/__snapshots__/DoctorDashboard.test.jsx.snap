// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoctorDashboard matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F8FAFC",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      {
        "color": "#64748B",
        "fontSize": 16,
      }
    }
  >
    Loading your dashboard...
  </Text>
</View>
`;
